nomos-libp2p/src/behaviour/nat/state_machine.rs

```diff
@@ -0,0 +1,749 @@
+            command_tx: command_tx.into(),
+        }
+    }
+
+    pub fn on_event<E>(&mut self, event: E)
+    where
+        E: TryInto<Event, Error = ()>,
+    {
+        let current_state = self.inner.take().expect("State to be Some");
+
+        match event.try_into() {
+            Err(()) => {
+                // Ignore unrecognized events
+                self.inner = Some(current_state);
+            }
+            Ok(event) => {
+                self.inner = Some(current_state.on_event(event, &self.command_tx));
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+struct State<S, E> {
+    pub state: S,
+    pub _phantom_event: PhantomData<E>,
+}
+
+#[derive(Debug, Clone)]
+struct CommandTx {
+    tx: UnboundedSender<Command>,
+}
+
+impl From<UnboundedSender<Command>> for CommandTx {
+    fn from(tx: UnboundedSender<Command>) -> Self {
+        Self { tx }
+    }
+}
+
+impl CommandTx {
+    fn send(&self, command: Command) {
+        self.tx.send(command).expect("Channel not to be closed");
+    }
+}
+
+#[cfg(not(test))]
+trait OnEvent<E>: Debug + Send {
+    fn on_event(self: Box<Self>, event: E, command_tx: &CommandTx) -> Box<dyn OnEvent<E>>;
+}
+
+#[cfg(test)]
+trait OnEvent<E>: Debug + Send + DynPartialEq {
+    fn on_event(self: Box<Self>, event: E, command_tx: &CommandTx) -> Box<dyn OnEvent<E>>;
+}
+
+#[cfg(test)]
+trait DynPartialEq {
+    fn box_eq(&self, other: &dyn std::any::Any) -> bool;
+    fn as_any(&self) -> &dyn std::any::Any;
+}
+
+impl<S, E> State<S, E> {
+    fn boxed<T, C>(self, next_state_ctor: C) -> Box<State<T, E>>
+    where
+        C: FnOnce(S) -> T,
+    {
+        let Self { state, .. } = self;
+        Box::new(State {
+            state: next_state_ctor(state),
+            _phantom_event: PhantomData,
+        })
+    }
+}
+
+impl OnEvent<Event> for State<Uninitialized, Event> {
+    fn on_event(self: Box<Self>, event: Event, _: &CommandTx) -> Box<dyn OnEvent<Event>> {
+        match event {
+            Event::NewExternalAddressCandidate(addr) => {
+                self.boxed(|state| state.into_test_if_public(addr))
+            }
+            _ => self,
+        }
+    }
+}
+
+impl OnEvent<Event> for State<TestIfPublic, Event> {
+    fn on_event(self: Box<Self>, event: Event, command_tx: &CommandTx) -> Box<dyn OnEvent<Event>> {
+        match event {
+            Event::ExternalAddressConfirmed(addr) if self.state.addr_to_test() == &addr => {
+                command_tx.send(Command::ScheduleAutonatClientTest(addr));
+                self.boxed(TestIfPublic::into_public)
+            }
+            Event::AutonatClientTestFailed(addr) if self.state.addr_to_test() == &addr => {
+                command_tx.send(Command::MapAddress(addr));
+                self.boxed(TestIfPublic::into_try_address_mapping)
+            }
+            Event::ExternalAddressConfirmed(addr) | Event::AutonatClientTestFailed(addr) => {
+                tracing::error!(
```

Comment:
Why is this an error? Could it be we try to test three different addresses? Or, as far as I can see, we can only test a single address?

Reply:
> That is wrong, but also potentially should never happen if we are always dealing with a single address. In that case we could even panic, not sure, or even better move to a different state that supports recovery.

is this transition invalid from the specification? If its an invalid transition then it should panic.